doctor-plugin/src/main/java/com/osacky/doctor/JavaHomeCheck.kt

```diff
@@ -3,44 +3,68 @@ package com.osacky.doctor
-                )
-            )
+        val extraMessage = extension.javaHomeHandler.extraMessage.get()
+        val failOnError = extension.javaHomeHandler.failOnError.get()
+
+        if (extension.javaHomeHandler.ensureJavaHomeIsSet.get() && environmentJavaHome == null) {
+            val message = buildString {
+                appendln("JAVA_HOME is not set.")
+                appendln("Please set JAVA_HOME so that switching between Android Studio and the terminal does not trigger a full rebuild.")
+                appendln("To set JAVA_HOME: (using bash)")
+                appendln("echo \"export JAVA_HOME=${'$'}(/usr/libexec/java_home)\" >> ~/.bash_profile")
+                appendln("or `~/.zshrc` if using zsh.")
+                extraMessage?.let {
+                    appendln()
+                    appendln(extraMessage)
+                }
+            }
+            val pill = pillBoxPrinter.createPill(message)
+            if (failOnError) {
+                throw GradleException(pill)
+            } else {
+                recordedErrors.add(pill)
+            }
         }
-        if (extension.ensureJavaHomeMatches.get() && !isGradleUsingJavaHome()) {
-            throw GradleException(
-                pillBoxPrinter.createPill(
-                    """
-                |Gradle is not using JAVA_HOME.
-                |JAVA_HOME is ${environmentJavaHome.toFile().toPath().toAbsolutePath()}
-                |Gradle is using ${gradleJavaHome.toPath().toAbsolutePath()}
-                |This can slow down your build significantly when switching from Android Studio to the terminal.
-                |To fix: Project Structure -> JDK Location.
-                |Set this to your JAVA_HOME.
-            """.trimMargin()
-                )
-            )
+        if (extension.javaHomeHandler.ensureJavaHomeMatches.get() && !isGradleUsingJavaHome()) {
+            val message = buildString {
+                appendln("Gradle is not using JAVA_HOME.")
+                appendln("JAVA_HOME is ${environmentJavaHome.toFile().toPath().toAbsolutePath()}")
+                appendln("Gradle is using ${gradleJavaHome.toPath().toAbsolutePath()}")
+                appendln("This can slow down your build significantly when switching from Android Studio to the terminal.")
+                appendln("To fix: Project Structure -> JDK Location.")
+                appendln("Set this to your JAVA_HOME.")
+                extraMessage?.let {
+                    appendln()
+                    appendln(extraMessage)
+                }
+            }
+            val pill = pillBoxPrinter.createPill(message)
+            if (failOnError) {
+                throw GradleException(pill)
+            } else {
+                recordedErrors.add(pill)
+            }
         }
     }
 
     override fun onFinish(): Finish {
+        if (recordedErrors.isNotEmpty()) {
+            logger.error(recordedErrors.joinToString("\n\n"))
```

Comment:
I have my own silly API design here that we'll need to follow in order to get a consistent printout of all the error messages together. A build can generate multiple warnings or errors so they are grouped using this API.

To get a consistent printout and so that all the errors are printed together at the end of the build we'll need to do something like this:
```suggestion
            return Finish.Message(recordedErrors.joinToString("\n\n"))
```
That being said, you might need to add a new class to `Finish` where you can pass a list of messages.

Here's an example from DownloadSpeedMeasurer: https://github.com/runningcode/gradle-doctor/blob/master/doctor-plugin/src/main/java/com/osacky/doctor/DownloadSpeedMeasurer.kt#L50